telemetry/data_feature_usage.go

```diff
@@ -33,32 +33,75 @@ type featureUsage struct {
-	ClusterIndex   *ClusterIndexUsage `json:"clusterIndex"`
-	TemporaryTable bool               `json:"temporaryTable"`
-	CTE            *m.CTEUsageCounter `json:"cte"`
-	CachedTable    bool               `json:"cachedTable"`
-	AutoCapture    bool               `json:"autoCapture"`
+	ClusterIndex         *ClusterIndexUsage    `json:"clusterIndex"`
+	TemporaryTable       bool                  `json:"temporaryTable"`
+	CTE                  *m.CTEUsageCounter    `json:"cte"`
+	CachedTable          bool                  `json:"cachedTable"`
+	AutoCapture          bool                  `json:"autoCapture"`
+	PlacementPolicyUsage *placementPolicyUsage `json:"placementPolicy"`
+}
+
+type placementPolicyUsage struct {
+	NumPlacementPolicies      uint64
+	NumDBWithPolicies         uint64
+	NumTableWithPolicies      uint64
+	NumPartitionWithPoliciess uint64
 }
 
 func getFeatureUsage(ctx sessionctx.Context) (*featureUsage, error) {
-	clusterIdxUsage, err := getClusterIndexUsageInfo(ctx)
+	var usage featureUsage
+	var err error
+	usage.ClusterIndex, err = getClusterIndexUsageInfo(ctx)
 	if err != nil {
 		logutil.BgLogger().Info(err.Error())
 		return nil, err
 	}
 
 	// transaction related feature
-	txnUsage := getTxnUsageInfo(ctx)
+	usage.Txn = getTxnUsageInfo(ctx)
 
-	// Avoid the circle dependency.
-	temporaryTable := ctx.(TemporaryOrCacheTableFeatureChecker).TemporaryTableExists()
+	usage.CTE = getCTEUsageInfo()
 
-	cteUsage := getCTEUsageInfo()
+	usage.AutoCapture = getAutoCaptureUsageInfo(ctx)
 
-	cachedTable := ctx.(TemporaryOrCacheTableFeatureChecker).CachedTableExists()
+	collectFeatureUsageFromInfoschema(ctx, &usage)
+	return &usage, nil
+}
 
-	enableAutoCapture := getAutoCaptureUsageInfo(ctx)
-	return &featureUsage{txnUsage, clusterIdxUsage, temporaryTable, cteUsage, cachedTable, enableAutoCapture}, nil
+// collectFeatureUsageFromInfoschema updates the usage for temporary table, cached table and placement policies.
+func collectFeatureUsageFromInfoschema(ctx sessionctx.Context, usage *featureUsage) {
+	if usage.PlacementPolicyUsage == nil {
+		usage.PlacementPolicyUsage = &placementPolicyUsage{}
+	}
+	is := GetDomainInfoSchema(ctx)
+	for _, dbInfo := range is.AllSchemas() {
+		if dbInfo.PlacementPolicyRef != nil {
+			usage.PlacementPolicyUsage.NumDBWithPolicies++
+		}
+
+		for _, tbInfo := range is.SchemaTables(dbInfo.Name) {
+			if tbInfo.Meta().TempTableType != model.TempTableNone {
+				usage.TemporaryTable = true
+			}
+			if tbInfo.Meta().TableCacheStatusType != model.TableCacheStatusDisable {
+				usage.CachedTable = true
+			}
+			if tbInfo.Meta().PlacementPolicyRef != nil {
+				usage.PlacementPolicyUsage.NumTableWithPolicies++
+			}
+			partitions := tbInfo.Meta().GetPartitionInfo()
+			if partitions == nil {
+				continue
+			}
+			for _, partitionInfo := range partitions.Definitions {
+				if partitionInfo.PlacementPolicyRef != nil {
+					usage.PlacementPolicyUsage.NumPartitionWithPoliciess++
```

Comment:
What is `NumPartitionWithPolicies` mean ? If a table has placement policy, all its partitions should have placement policies.

Reply:
I've updated it to `NumPartitionWithExplicitPolicies` with a comment.